### PR TITLE
feat(gatsby): lazy bundle page components in dev server

### PR DESCRIPTION
**Pull request by KyleAMathews**
This means we only compile the core runtime code initially — page components are only compiled when the user visits the page.

This can lead to time savings on compiling the dev bundle up to 80% — e.g. building the dev bundle for one test site w/ a large number of components dropped from 42 seconds to 5 seconds.

On gatsbyjs.com, the bundle compilation time dropped from 145 seconds to 27 seconds for a 81% decrease. Smaller sites won't see as a dramatic of difference — more around ~40-60%.

### CODE REVIEW

File: packages/gatsby-link/src/index.js
Comment (DSchau) [inline packages/gatsby-link/src/index.js:210]:
```suggestion
```

File: packages/gatsby/cache-dir/dev-loader.js
Comment (pvdz) [inline packages/gatsby/cache-dir/dev-loader.js:17]: This promise doesn't reject (errors would be lost) and the timeout timer does not call reject nor resolve, meaning I think it ends up in a zombie state once it hits that path, since the interval that would call `resolve` and is the only way to resolve the promise is stopped.

File: packages/gatsby/cache-dir/loader.js
Comment (pvdz) [inline packages/gatsby/cache-dir/loader.js:239]: What happens when `loadComponent` throws an error? Or when the `then` callback does? I think this goes to most promises here. Is there a general uncaught rejection handler in place?

File: packages/gatsby/src/bootstrap/requires-writer.ts
Comment (pvdz) [inline packages/gatsby/src/bootstrap/requires-writer.ts:191]: Scaling red flag. What are `n` and `m` here? What's the worst case runtime? It's fine if it's tiny but if it scales with the number of pages/nodes/content creating a `Set` first might be worth it.

File: packages/gatsby/src/redux/actions/public.js
Comment (pvdz) [inline packages/gatsby/src/redux/actions/public.js:1400]: Mismatch?

**Review by pvdz — COMMENTED** [review]
(some comments, by no means a thorough review)

**Comment by pvdz** [comment]
You had me at 80%! :rocket:

File: packages/gatsby/src/bootstrap/requires-writer.ts
Comment (KyleAMathews) [reply packages/gatsby/src/bootstrap/requires-writer.ts:190]: This is looping over page components which grows slowly even for very large sites (since they'll have most of their pages getting built by one page component). It's not the number of pages but the number of _page types_. I tested this a lot with a 1000 unique page components & didn't notice any slow downs.

File: packages/gatsby/src/redux/actions/public.js
Comment (KyleAMathews) [reply packages/gatsby/src/redux/actions/public.js:1400]: with what? Not following

Comment (KyleAMathews) [reply packages/gatsby/src/redux/actions/public.js]: oh the params. yeah

File: packages/gatsby/cache-dir/loader.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/loader.js]: I don't think so. I know we've talked about adding a general error boundary for this stuff in React land but that work hasn't been done yet AFAIK. @wardpeet or @pieh might know.

File: packages/gatsby/cache-dir/dev-loader.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/dev-loader.js:17]: There's nothing here to reject on I don't think. We're just telling the server to start compiling the page component and then... waiting. We're not doing anything per se.

File: packages/gatsby/cache-dir/loader.js
Comment (pieh) [reply packages/gatsby/cache-dir/loader.js]: ProdLoader's `loadComponent` method current implementation catches rejection internally and it resolve to `null` - https://github.com/gatsbyjs/gatsby/blob/ae2c3dab9fe6b26a968b7762892243914ee10a6b/packages/gatsby/cache-dir/loader.js#L439-L447

There should be some `null` checks by whatever call `loadComponent` method, and I think behaviour should be similar as when user tries to navigate to a page that can't load `page-data.json` file - https://github.com/gatsbyjs/gatsby/blob/ae2c3dab9fe6b26a968b7762892243914ee10a6b/packages/gatsby/cache-dir/loader.js#L476-L493

(above code is done for the most part as resource loading resilience attempt, to try to load the page in traditional way if resources are blocked or missing)

DevLoader (speaking about `master` branch, not this PR as I didn't follow changes yet) doesn't have anything like that  but it also doesn't do any async work, because we don't split chunks in develop and everything load at once, so if "module loading" would cause errors it will likely break entire runtime at the time we import `sync-requires` (so https://github.com/gatsbyjs/gatsby/blob/ae2c3dab9fe6b26a968b7762892243914ee10a6b/packages/gatsby/cache-dir/app.js#L11 )?

**Comment by KyleAMathews** [comment]
Guessing the development_runtime failures are from the compilation not getting triggered under certain scenarios or hitting timeouts. Will dig in Monday.

File: packages/gatsby/src/bootstrap/requires-writer.ts
Comment (wardpeet) [inline packages/gatsby/src/bootstrap/requires-writer.ts:190]: let's go for better variable names
```suggestion
  const cleanedClientVisitedPageComponents = components.filter(component =>
    clientVisitedPageComponents.some(pageComponentChunkName => pageComponentChunkName === component.componentChunkName)
  )
```

File: packages/gatsby/cache-dir/dev-loader.js
Comment (wardpeet) [inline packages/gatsby/cache-dir/dev-loader.js:36]: setInterval can put lots of pressure on the event loop when the browser can't keep up. setTimeout relieves pressure.

```suggestion
          const checkForUpdates = () => {
            if (process.env.NODE_ENV !== `test`) {
              delete require.cache[
                require.resolve(`$virtual/lazy-client-sync-requires`)
              ]
            }
            const lazyRequires = require(`$virtual/lazy-client-sync-requires`)
            if (lazyRequires.lazyComponents[chunkName]) {
              return resolve(lazyRequires.lazyComponents[chunkName])
            }
            
            setTimeout(checkForUpdates, 100)
          }
          setTimeout(checkForUpdates, 100)
```

**Review by wardpeet — COMMENTED** [review]
Awesome! I'm not the biggest fan of the redux dance, I rather have it all scoped inside a webpack plugin but that would be more work so let's go with this approach 💪 💯 

What's the delay on page load?

File: packages/gatsby/cache-dir/dev-loader.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/dev-loader.js:36]: 👍

**Comment by KyleAMathews** [comment]
>What's the delay on page load?

depends on how big the page component is but for simple sites it was very reasonable e.g. < 200ms

**Comment by KyleAMathews** [comment]
Ended up going on quite the journey today — pages with static queries weren't loading — turns out we don't know what the static queries for a page component are until we bundle it. So we have to first trigger bundling it, wait until it's written out, and then finally load it in the browser.

https://github.com/gatsbyjs/gatsby/pull/27884/commits/9fd03eaeb1860a1dbe0f587cb06db1cf7215afce

**Comment by KyleAMathews** [comment]
This is definitely weirder now so not sure if there's a better way to handle this. With the per-page querying work, it'd probably be easier to track this as we'll be tracking the state of each page cc @pieh @vladar

**Comment by pieh** [comment]
> With the per-page querying work, it'd probably be easier to track this as we'll be tracking the state of each page

We can potentially trigger that in `page-data` express handler ( https://github.com/gatsbyjs/gatsby/pull/27882 - that will be one of triggers for query on demand). 

Tricky part however is that in legacy mode (non query on demand) `page-data` endpoints will be hit for preloading too and this would need additional work so we can distinguish between `page-data` fetches needed to render a page from those that preload data for other pages. Unless we would want to precompile those on hovers as well? Would definitely need some testing

**Comment by KyleAMathews** [comment]
>Unless we would want to precompile those on hovers as well

yeah I think this is fine — probably preferable even as it lowers latency a bit to seeing the new page. Hovering is a strong signal that the user is going to that page.

**Comment by KyleAMathews** [comment]
>We can potentially trigger that in page-data express handler ( #27882 - that will be one of triggers for query on demand).

yeah that'd be perfect — conceptually it's "do this bucket of work to make this page ready for development" — which includes running the query and with this PR, adding its page component to the dev bundle.

**Comment by wardpeet** [comment]
I'm still not sold on the solution. We touch many files and static-query & page-query data is already pretty brittle. We should do proper testing on production and develop with react-refresh, hmr, ...

I'm more confident if we move to what we do in production with dynamic imports. We could leverage https://github.com/liximomo/lazy-compile-webpack-plugin to lazy compile those imports so it should be pretty fast as well. 
1) Complexity drops because we only need changes in webpack
2) We should see the same speed benefits
3) It improves bundling and viewing develop all together

I'm not sure if static-queries extraction would still work

**Comment by KyleAMathews** [comment]
https://github.com/liximomo/lazy-compile-webpack-plugin is a nice find and normally I'd prefer something like that as well — for our particular use case though I don't think it works as we're waiting on two things: the new component to be bundled and loaded into the client (which this plugin handles) and for the static queries for a page component to be discovered and run — https://github.com/gatsbyjs/gatsby/pull/27884/#issuecomment-724429471

Because of the latter, the backend also needs to know what pages the user has visited so it can inform the client that the static queries have now been added so it's safe to render. 

So it's a bit more work to do ourselves what the plugin does but we need the lower-level control.

This wouldn't affect HMR/fast-refresh. I liken its effect to living in a giant mansion (say a Gatsby Mansion 😆 ). Currently when you enter, we turn on all the lights and heat in every room in the mansion, regardless of how obscure. This changes things so we only turn on the lights/heat when you enter a room. As far as webpack is concerned, things work exactly the same — it's just working with a lot less code. The main thing the user will notice is HMR should be a lot faster on large sites.

How would this affect production? I was careful to not touch any prod code pathways — what do you think needs tested?

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (pieh) [inline packages/gatsby/cache-dir/gatsby-browser-entry.js:64]: Why fallback is needed here?

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/gatsby-browser-entry.js:64]: it probably doesn't — I think I added it when I was trying to debug something and it stayed. Removing.

File: packages/gatsby/src/utils/page-data.ts
Comment (pieh) [inline packages/gatsby/src/utils/page-data.ts:98]: This doesn't guarantee that `notInDevBundle` will not end up in production artefacts, as we do reuse `page-data.json` written in develop - if you run `gatsby develop`, stop it after it's done and run `gatsby build` afterwards you will likely see this field there

File: packages/gatsby/cache-dir/loader.js
Comment (pieh) [inline packages/gatsby/cache-dir/loader.js:158]: Can we instead of adding `notInDevBundle` field, have the syncRequires have all the keys that it would have normally, just instead of importing actual page component, have something like `Symbol("not-yet-compiled")` and check for that and that would be trigger to tell dev server to compile it? That would not "taint" the `page-data` files with dev-only metadata

Not sure if `Symbol` would work here nicely, but even some hardcoded string could be used instead

File: packages/gatsby/src/utils/page-data.ts
Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts:98]: yeah — I realize this. I consider this acceptable as:

1. it wouldn't affect the production site and it's a tiny amount of extra data if it does end up in production page-data.json files
2. most people have a separate build server so wouldn't run into this scenario

If there's a different way we can think of to pass the state of whether a path's page component is in the bundle or not that'd be great. E.g. I thought of exposing a route for the runtime to query for this. But honestly this feels like perhaps the cleanest. It ties together writing out the staticQueryHashes w/ changing this boolean. Any other solution would have race conditions.

File: packages/gatsby/cache-dir/loader.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/loader.js:158]: ahhh yeah I think this would work! I'll try this

File: packages/gatsby/cache-dir/dev-loader.js
Comment (pvdz) [reply packages/gatsby/cache-dir/dev-loader.js]: Anything can crash. `this.loadComponent` can fail for any number of reasons :shrug:

**Comment by pieh** [comment]
initial NIT for env var name:
`GATSBY_EXPERIMENT_LAZY_DEVJS`

we currently have couple with `GATSBY_EXPERIMENTAL_` prefix ( vs `GATSBY_EXPERIMENT_`):
 - `GATSBY_EXPERIMENTAL_PAGE_BUILD_ON_DATA_CHANGES`
 - `GATSBY_EXPERIMENTAL_QUERY_CONCURRENCY`
 - `GATSBY_EXPERIMENTAL_DISABLE_SCHEMA_REBUILD`
 - `GATSBY_EXPERIMENTAL_ENABLE_ADMIN`

(and I also used this convention when using `GATSBY_EXPERIMENTAL_QUERY_ON_DEMAND`)

File: packages/gatsby/cache-dir/app.js
Comment (pieh) [inline packages/gatsby/cache-dir/app.js:38]: Is it ok to keep this for non `GATSBY_EXPERIMENT_LAZY_DEVJS` runs?

Can it actually help in regular `gatsby develop`? If not maybe this should be just in `if (process.env.GATSBY_EXPERIMENT_LAZY_DEVJS) {` block?

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/app.js:38]: yeah this is a general fix so I didn't think it needed to be behind a flag

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (pieh) [inline packages/gatsby/cache-dir/gatsby-browser-entry.js:28]: I think this shouldn't be needed anymore since we merge 2 separate caches (loader and socket.io), but also this is fine to keep for now especially behind the flag, but we can mark this to check this in follow up?

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (pieh) [inline packages/gatsby/cache-dir/gatsby-browser-entry.js:95]: Same as https://github.com/gatsbyjs/gatsby/pull/27884/files#r525431628

File: packages/gatsby/cache-dir/loader.js
Comment (pieh) [inline packages/gatsby/cache-dir/loader.js:550]: This is not tree shakable and we already have https://github.com/gatsbyjs/gatsby/blob/6858f22fc9a64f71756d5cd5122fab99d12749cf/packages/gatsby/cache-dir/loader.js#L540-L542

that does the same

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/gatsby-browser-entry.js:28]: yeah that makes sense — so when I remove this, we can just use the context?

File: packages/gatsby/cache-dir/app.js
Comment (vladar) [inline packages/gatsby/cache-dir/app.js:133]: Can you elaborate a bit on how it can bypass the `loadPage` promise? Not entirely clear from this comment.

File: packages/gatsby/cache-dir/loader.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/loader.js:550]: yeah, not sure why I added this now — I thought it didn't work for some reason (or maybe I just didn't notice it).

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (pieh) [reply packages/gatsby/cache-dir/gatsby-browser-entry.js:28]: You should be yes, because https://github.com/gatsbyjs/gatsby/blob/6858f22fc9a64f71756d5cd5122fab99d12749cf/packages/gatsby/cache-dir/query-result-store.js#L94-L131 has the context provider and value is pulled from `loader` on updates (either toward the end of `loadPage` execution or on hot data updates)

But I didn't run it locally without it, so hard to say - so that's why I think it might be ok to keep it for now to minimize amount of changes needed to get this in and streamline afterward if it turns out to not be needed?

**Review by vladar — COMMENTED** [review]
I did a shallow review - mostly checked if there is something that leaks through the experimental flag. Couldn't find anything particularly standing out.

Only the additional `loader.loadPage(window.location.pathname).then()` in `app.js is not quite clear.

File: packages/gatsby/src/utils/start-server.ts
Comment (pieh) [inline packages/gatsby/src/utils/start-server.ts:129]: can we add `bodyParser.json()` middleware here instead of adding global one? ( I don't think adding global one will break things, it just seems wasteful for all the other requests?)
```suggestion
    app.post(`/___client-page-visited`, bodyParser.json(), (req, res, next) => {
```

We use this per request handler middleware already in https://github.com/gatsbyjs/gatsby/blob/6858f22fc9a64f71756d5cd5122fab99d12749cf/packages/gatsby/src/utils/telemetry-server.ts#L44

File: packages/gatsby/cache-dir/gatsby-browser-entry.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/gatsby-browser-entry.js]: pushed

File: packages/gatsby/cache-dir/app.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/app.js:133]: not entirely sure about this now — lemme double-check

File: packages/gatsby/src/utils/start-server.ts
Comment (KyleAMathews) [reply packages/gatsby/src/utils/start-server.ts:129]: 👍  makes sense

File: packages/gatsby/cache-dir/app.js
Comment (KyleAMathews) [reply packages/gatsby/cache-dir/app.js]: fake news. Removed.

File: packages/gatsby/cache-dir/__tests__/dev-loader-lazy.js
Comment (pieh) [inline packages/gatsby/cache-dir/__tests__/dev-loader-lazy.js:550]: Just comparing those tests to current `dev-loader.js` it seems like it has only minor adjustments for setup (setting `GATSBY_EXPERIMENT_LAZY_DEVJS` env var).

Changing the field name in `createSyncRequires` helper and ... removes one test:

```
    it(`should return an error when component cannot be loaded`, async () => {
      const syncRequires = createSyncRequires({
        chunk: false,
      })
      const devLoader = new DevLoader(syncRequires, [])
      const pageData = {
        path: `/mypage/`,
        componentChunkName: `chunk`,
        staticQueryHashes: [],
      }
      devLoader.loadPageDataJson = jest.fn(() =>
        Promise.resolve({
          payload: pageData,
          status: `success`,
        })
      )

      await devLoader.loadPage(`/mypage/`)
      const expectation = devLoader.pageDb.get(`/mypage`)
      expect(expectation).toHaveProperty(`status`, `error`)
      expect(emitter.emit).toHaveBeenCalledTimes(0)
    })
```

I'm not sure of the value of that test to begin with but it looks a bit look a smell that something doesn't behave the same?

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/__tests__/dev-loader-lazy.js:550]: This is an intended change — w/ lazy bundling — it isn't easy to know if a page component can't be loaded. Perhaps we could add a timeout or something but it didn't seem like this made sense to test against.

File: packages/gatsby/cache-dir/ensure-page-component-in-bundle.js
Comment (pieh) [inline packages/gatsby/cache-dir/ensure-page-component-in-bundle.js:32]: do we need to consider that requested `chunkname` actually would never show up in the bundle and this would never resolve or reject?

because of a lot of async execution here (including networking) there still might be the cases where template might no longer be there (was removed since initial request)

We did hit problem like this Vlad when working on query on demand in this test scenario: https://github.com/gatsbyjs/gatsby/blob/6858f22fc9a64f71756d5cd5122fab99d12749cf/e2e-tests/development-runtime/cypress/integration/navigation/redirect.js#L81-L112

which removes the page and immediately tries to visit it so page is still in store ( we didn't receive event from chokidar yet to remove the page), then pages get removed while we do wait for `page-data` for it - I think this is quite similar case?

Comment (pieh) [reply packages/gatsby/cache-dir/ensure-page-component-in-bundle.js:32]: I think it's not blocking merge because this is experimental - but something to keep in mind?

Comment (KyleAMathews) [reply packages/gatsby/cache-dir/ensure-page-component-in-bundle.js:32]: yeah... actually timing out makes a lot of sense — we can check for say 5 seconds and then quit. Otherwise the runtime will just keep looping forever trying to load it which is wasteful at best and could slow things down.

**Review by pieh — APPROVED** [review]
This looks good to get into master branch

it might have some edge cases due to timing/async, but all of this is behind experimental flag so I think it's good to go!

faster `gatsby develop`, here we come :)

**Review by pieh — APPROVED** [review]
re-approve after merging master

**Pull request merged**
